mp2-v1/src/values_extraction/api.rs

```diff
@@ -30,30 +33,53 @@ use recursion_framework::{
 use serde::{Deserialize, Serialize};
 use std::array;
 
-type LeafSingleWire = LeafSingleWires<MAX_LEAF_NODE_LEN>;
-type LeafMappingWire = LeafMappingWires<MAX_LEAF_NODE_LEN>;
 type ExtensionInput = ProofInputSerialized<InputNode>;
 type BranchInput = ProofInputSerialized<InputNode>;
 const NUM_IO: usize = PublicInputs::<F>::TOTAL_LEN;
 
 /// CircuitInput is a wrapper around the different specialized circuits that can
 /// be used to prove a MPT node recursively.
 #[derive(Serialize, Deserialize)]
-pub enum CircuitInput {
-    LeafSingle(LeafSingleCircuit<MAX_LEAF_NODE_LEN>),
-    LeafMapping(LeafMappingCircuit<MAX_LEAF_NODE_LEN>),
+pub enum CircuitInput<
+    const NODE_LEN: usize,
+    const MAX_COLUMNS: usize,
+    const MAX_FIELD_PER_EVM: usize,
+> where
+    [(); PAD_LEN(NODE_LEN)]:,
+{
+    LeafSingle(LeafSingleCircuit<NODE_LEN, MAX_COLUMNS, MAX_FIELD_PER_EVM>),
+    LeafMapping(LeafMappingCircuit<NODE_LEN, MAX_COLUMNS, MAX_FIELD_PER_EVM>),
+    LeafMappingOfMappings(LeafMappingOfMappingsCircuit<NODE_LEN, MAX_COLUMNS, MAX_FIELD_PER_EVM>),
     Extension(ExtensionInput),
-    BranchSingle(BranchInput),
-    BranchMapping(BranchInput),
+    Branch(BranchInput),
 }
 
-impl CircuitInput {
+impl<const NODE_LEN: usize, const MAX_COLUMNS: usize, const MAX_FIELD_PER_EVM: usize>
+    CircuitInput<NODE_LEN, MAX_COLUMNS, MAX_FIELD_PER_EVM>
+where
+    [(); PAD_LEN(NODE_LEN)]:,
+{
     /// Create a circuit input for proving a leaf MPT node of single variable.
-    pub fn new_single_variable_leaf(node: Vec<u8>, slot: u8, column_id: u64) -> Self {
+    pub fn new_single_variable_leaf(
+        node: Vec<u8>,
+        slot: u8,
+        evm_word: u32,
+        num_actual_columns: usize,
+        extracted_column_identifiers: &[F],
+        table_info: Vec<ColumnInfo>,
```

Comment:
Wouldn't be simpler to take as input a `MetadataGadget` for all these methods? In the integration code it seems to me you always construct first a `MetadataGadget` instance to be placed in `StorageSlotInfo`, and then you decompose it to provide these input values separately. So I am wondering if it makes more sense to take an already built `MetadataGagdet` instance as input here. Wdyt? Ofc this would be applicable to all the subsequent initialization methods for this struct.